scripts/rouleth/rouleth.ts

```diff
@@ -0,0 +1,210 @@
+    const addressClean: string = address.replace(/\s/g, '');
+    if (ethers.utils.isAddress(addressClean)) {
+      if (await isContract(addressClean)) {
+        console.error('contract are not eligible:\t', addressClean);
+        return false;
+      }
+      return addressClean;
+    } else {
+      try {
+        const resolvedEns = await provider.resolveName(addressClean);
+        if (resolvedEns !== null && address !== '.eth') {
+          return resolvedEns;
+        }
+        console.error('failed to resolve ens:\t\t', addressClean);
+        return false;
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  };
+
+  // check all address are valid, and await resolution
+  const promiseResolver = await Promise.all(
+    address.map((add) => verifyAddress(add))
+  );
+
+  // insure we only have string in the array and false / undefined values from address check are removed
+  const addressFromBack: string[] = promiseResolver.filter(
+    (val) => typeof val === 'string'
+  ) as string[];
+
+  return addressFromBack;
+};
+
+const lottery = async (
+  addBack: Array<string>,
+  addGraph: Array<LandOwner>,
+  maxWinnerNb: number,
+  blockNumber: number
+) => {
+  const seed: string = (await ethers.provider.getBlock(blockNumber)).hash;
+
+  class SeededRand {
+    private seed;
+    constructor(seed: string) {
+      this.seed = seed;
+    }
+
+    nextInteger(): number {
+      this.seed = solidityKeccak256(['bytes32'], [this.seed]);
+      return parseInt(this.seed, 16) % Math.floor(Math.random() * 100);
+      // return BigNumber.from(this.seed).mod(Number.MAX_SAFE_INTEGER).toNumber();
+    }
+  }
+
+  // return an array of address subscribe on the back
+  // if an address own x land, add x time the address to the array (x >= 1)
+  const extendedAddressArray = async () => {
+    const extendedArray: Array<string> = [];
+    addBack.filter((itemAddress: string) => {
+      extendedArray.push(itemAddress);
+      addGraph.map((itemLandOwner: LandOwner) => {
+        if (itemLandOwner.id === itemAddress) {
+          for (let i = 0; i < itemLandOwner.nb; i++) {
+            extendedArray.push(itemLandOwner.id);
+          }
+        }
+      });
+    });
+    return extendedArray;
+  };
+
+  // return an array of max winner address randomly selected in the extendedAddressArray
+  const randomWinnersSelection = (addresses: string[], seed: string) => {
+    const result: Array<string> = [];
+    const rand = new SeededRand(seed);
+    const sortedList = addresses.sort((address1, address2) =>
+      BigNumber.from(address1).sub(BigNumber.from(address2)).gt(0) ? 1 : -1
+    );
+    for (let i = 0; i < maxWinnerNb; i++) {
+      const index = rand.nextInteger() % sortedList.length;
+      const address = sortedList[index];
+      result.push(address);
+      sortedList.splice(index, 1);
+      // sortedList = sortedList.filter((v) => v.toLowerCase() !== address.toLowerCase()) // If only 1 winning ticket per address
```

Comment:
as discussed, pending confirmation the commented line should be what we use to not have someone win multiple time.
I ll get back on this. Else contact Buche about it